### DISCUSSION TITLE
Turning on Holtzlag-Boville (HB) where CLUBB is not active

### DISCUSSION THREAD

**Discussion by PeterHjortLauritzen** [NCAR/amwg_dev#284] [TF1: Unified physics for L58/L93]
It has been found difficult to stabilize the spectral-element dynamical core (and other dycores such as MPAS) when the model lid is above ~80km. In the case of the spectral-element dynamical core excessively small time-steps and excessively large hyperviscosity coefficients are used to keep the model stable. Hence it runs very inefficiently!

@hanli-liu suggested turning on the HB boundary layer scheme that performs Richardson number based mixing in the free troposphere and above (@adamrher implemented a "hacked" version a while ago). Doing so makes WACCM much more stable. See a more detailed discussion on time-steps and viscosity settings here:

https://github.com/PeterHjortLauritzen/CAM/wiki/stability-with-spectral-element-dynamical-core 

This discussion is to discuss and evaluate the consequences of the "Holtzlag-Boville (HB) where CLUBB is not active" configuration of CAM7.

Here are results from a 1 year WACCM ne16pg3 (~2 degree)  simulation (note that the y-axis are different):

Vertical profile of KVH and KVH_CLUBB at the Equator and lon=62.5 (1 year average):

<img width="664" alt="Screen Shot 2023-05-30 at 2 55 06 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/b84ccf92-731b-450b-bc15-67db3ba1ae5a">

Same plot on linear scale:

<img width="659" alt="Screen Shot 2023-05-30 at 4 27 12 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/2dec8c2c-0b25-4206-94bc-a90727e76043">

2D plots for (left) KVH at 123hPa and (right) KVH_CLUBB at 176hPa:

<img width="1722" alt="Screen Shot 2023-05-30 at 4 34 02 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/7862d332-17c2-4a60-9d3b-48c65925433a">

Note difference in scales (left and right plot), i.e. HB does a lot less mixing (factor 30ish) compared to CLUBB (and always above CLUBB - by design)

@JulioTBacmeister @islasimpson @adamrher @vlarson

**Comment by rgarc1a** [comment]
Good discussion. 

"It has been found difficult to stabilize the spectral-element dynamical core". Indeed, as I have been mentioning now for quite a while.

A few comments:

1. HB is meant to address dynamical instabilities of the flow in the free atmosphere. Resolved waves will grow as exp(z/2H) unless they are dissipated. Since HB is a Ri-dependent diffusivity, it operates only when the resolved flow is about to become locally unstable; 
2. CLUBB, as far as I can tell, was never meant to address this type of instability. Thus, it is not surprising that a CLUBB-only model becomes unstable frequently at higher altitudes;
3. Trying to control the flow instability by increasing the number of dynamical sub-steps and relying on hyper-diffusion is spectacularly inefficient, as Peter notes;
4. Because HB only turns on locally where the flow is about to become dynamically unstable, which occurs sporadically, the global-mean HB diffusivities are much smaller than what CLUBB produces;
5. In general, CLUBB operates in the troposphere and HB above the tropopause; but there is an exception, the UTLS, where both operate, presumably for different reasons. In the case of HB, the reason is likely to be dynamical instabilities around the subtropical jets (and maybe also the descending QBO?) Thus, there is an argument to be made for letting both parameterizations operate throughout the model domain. This would not be double-counting because each parameterization is addressing a different type of instability.

As a side comment, the first plot of Peter's entry illustrates why 0.01 hPa is such a bad place to locate an upper boundary (as is currently the case in the MT model). And that is just the resolved dynamics; the same applies to the parameterized GW drag.  Defining a reasonable UBC for chemistry is just as problematic. Too many things are changing too rapidly at 0.01 hPa, and these things depend on what is happening outside the model domain, i.e.,  above 0.01 hPa. An UBC near ~0.001 hPa, which is higher than MT but somewhat lower than WACCM, would solve these problems nicely in a physically meaningful way. It would also make the simulation credible throughout the mesosphere, something that the MT model, which is only marginally cheaper, does not do. We are talking about maybe 10-12 extra levels at 1/2 scale height (3.5 km resolution), or about a 12% premium over MT.

**Comment by adamrher** [comment]
Three main points in response to this thread.

(1) One should not confuse KVH_CLUBB for how moisture / heat are handled by clubb turbulence; KVH_CLUBB is only applied to tracers, whereas turbulent fluxes of heat and moisture are solved prognostically. Julio, however, tells he he thinks KVH_CLUBB might not be a terrible surrogate for overall CLUBB behavior, especially for momentum diffusion. Although we are running with prognostic momentum going forward, and I'm not sure I agree that KVH_CLUBB bears resemblance to -d/dz (prognostic clubb fluxes).

(2) CLUBB does not operate above 1 hPa; KVH_CLUBB is zero above 1 hPa. You cannot infer how CLUBB would behave above 1 hPa from the KVH_CLUBB plots above. When we have lifted the 1 hPa limit in the past, the model crashed. I admittedly have not looked into why this is the case.

(3) I don't think I agree that HB and CLUBB are addressing different types of instabilities; I would argue that HB is more responsive to Ri instability by design; whereas CLUBB tries to incorporate the impact of buoyancy and sheer from the ground up by solving the turbulent fluxes prognostically.

**Comment by rgarc1a** [comment]
OK.

As regards, 

"I don't think I agree that HB and CLUBB are addressing different types of instabilities; I would argue that HB is more responsive to Ri instability by design; whereas CLUBB tries to incorporate the impact of buoyancy and shear”

the fact remains that the model with HB can run where without it it crashes. Eppur si muove.

Rolando Garcia
NCAR/ACOM
P.O. Box 3000
Boulder, CO 80307-3000
(303) 497-1446
***@***.***

**Reply by adamrher** [reply]
I am in agreement with using the HB scheme to stabilize our mid and high top models, but I'm only comfortable with it being active **above** where clubb is active, as Peter currently has it implemented.

Normally, PBL schemes provide diffusion all the way to top of the model, but clubb is not normal and is not doing this. So this HB extension also makes sense from a historical GCM perspective. That big spike in HB around 0.001 hPa is a compelling anecdote for why you shouldn't neglect vertical mixing up there.

**Reply by cecilehannay** [reply]
@adamrher: I am ready to start a MT run (see Simone's request at   https://github.com/NCAR/amwg_dev/issues/281)
She wants to include the HB as Peter originally formulated. I assume I can go with that. Thanks.

**Reply by rgarc1a** [reply]
This would be very interesting. 

The questions one would like to answer are: Can the SE model run without crashing at reasonable nsplit with the HB parameterization? and does HB interfere with anything CLUBB might be doing?

We can’t answer the second question without we try it. Maybe it does not matter? I am not sure: there is a peak in HB Kvh that appears to partly overlaps what CLUBB is doing in the upper troposphere.

The big spike in Kvh at ~0.001 hPa is there because this is where resolved waves become large and want to break. HB mimics the effect of wave breaking, which cannot be simulated at all in a hydrostatic model. Parameterized GW also want to break at these altitudes, starting maybe a little bit lower. In fact, some of the resolved waves that are breaking at 0.001 hPa may have been excited by parameterized GW accelerations lower down.

By 0.0001 hPa all of this is done. The reason why an UBC condition would be natural/easy to implement there. Note that 0.0001 hPa is what I meant to write above (not 0.001 hPa).

**Comment by PeterHjortLauritzen** [comment]
In the last CAM7 meeting we discussed the HB implementation which has a free atmosphere component and a PBL component. @tto061 provide code template 

https://github.com/ESCOMP/CAM/issues/846#issuecomment-1604752567

to only turn on the free atmospheric component that @PeterHjortLauritzen has implemented in a current code base.

Two experiments were performed: one in which the free atmosphere HB is turned on through-out the atmosphere and one where it is only on where CLUBB is not active. WACCM 2 degree with 135 levels was run for 1 month. Below is 1 month average of (left) KVH and (right) KVH_CLUBB  at 995hPa (nlev-1); color scale is 0-60 on both plots:

![Screen Shot 2023-06-28 at 9 23 42 AM](https://github.com/NCAR/amwg_dev/assets/8782989/5f9cd864-0e7e-44c8-b570-7cfaddeed9e8)

Clearly HB (even without the PBL component) is very active.

Same plots but one level higher up (987hPa) CLUBB takes over:

![Screen Shot 2023-06-28 at 9 26 20 AM](https://github.com/NCAR/amwg_dev/assets/8782989/3606b581-99a8-40c2-b988-be06dd109632)

Vertical profile of KVH (over some rather active area):

![Screen Shot 2023-06-28 at 9 34 23 AM](https://github.com/NCAR/amwg_dev/assets/8782989/13eff510-9cb5-4427-9e56-f909c990bb29)

Vertical profile of KVH_CLUBB (same point as plot above); note that y-axis range is 2x compared to plot above):

![Screen Shot 2023-06-28 at 9 34 55 AM](https://github.com/NCAR/amwg_dev/assets/8782989/f00bc4f5-6403-475a-a821-93e6756618ff)

These (preliminary) results (though short in duration and limited in scope) do indicate that there is a risk of HB interferring with CLUBB in the boundary layer which may not be desirable (as mentioned by @adamrher above). Hence I propose to use this new implementation of turning on only the free atmosphere component of HB where CLUBB is not active ...

**Reply by adamrher** [reply]
why are the PBL values in KVH so much larger than the plot at the top of this thread? Was KVH just not computed in that earlier run, where clubb is active?

**Reply by PeterHjortLauritzen** [reply]
Different location and in the plots above HB was not active where CLUBB is active.

**Comment by vlarson** [comment]
Does anyone know what happens when CLUBB is turned on for all vertical levels up to the model top, and HB is shut off for all vertical levels?  I am guessing that the model crashes.  Does anyone know what are the symptoms of the crashes?

**Reply by PeterHjortLauritzen** [reply]
I tried that in WACCM (top around 140km) and the model crashed immediately. Don't know why! (no comprehensible error messages)

**Reply by vlarson** [reply]
If the model crashes immediately, then I wonder if CLUBB's saturation mixing ratio can't handle the low pressure, density, or temperature and produces a NaN.

**Reply by adamrher** [reply]
That's an interesting lead; very plausible. How would you suggest we modify the pdf closure above the current ~1 hPa "lid" on moist physics?

**Reply by vlarson** [reply]
I'm not sure where the bug lies.  I glanced at the code but did not find an obvious culprit.  Is there an easy way to figure out which line of code is causing the crash?  Or is inserting print statements the best way right now?

**Reply by PeterHjortLauritzen** [reply]
I can try a run with WACCM (2 degree) with debug turned on.

**Reply by vlarson** [reply]
Thanks, Peter.  Usually if the model crashes on the first time step, it is possible (with some digging) to find the source of the crash.

**Reply by PeterHjortLauritzen** [reply]
FYI: it worked with a top at ~80km (https://github.com/NCAR/amwg_dev/discussions/205)
FYI2: I mispoke about no error messages (at least in my new experiment)

My setup is WACCM (top is at 0.0002Pa) at 2 degrees and 135 levels (Nick Davis setup) with

trop_cloud_top_press = 0.0002D0

This version runs 2 time-steps and blows up in the dynamical core in level 1:

155:k,dp_star_moist,u,v,T    1  -0.000000480596   -447.50   -101.91    398.03

Looking at 'STEND_CLUBB','RVMTEND_CLUBB','UTEND_CLUBB','VTEND_CLUBB' it is STEND_CLUBB that looks crazy:

In level 2 the values are over 500 right before blow-up:

<img width="573" alt="Screen Shot 2023-06-29 at 3 46 53 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/26f1ce36-98bb-4488-a25f-f851aa8b7a27">

"normal" values seem less than 3ish (below is bottom most level; surface)

<img width="862" alt="Screen Shot 2023-06-29 at 3 48 20 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/45fe6cce-4183-4351-908e-96b5e9208ff5">

UTEND_CLUBB and VTEND_CLUBB look reasonable (level 5 shown below):

<img width="857" alt="Screen Shot 2023-06-29 at 3 50 23 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/08ff6df8-a55e-4a9a-9faf-438009bd0f87">

**Reply by adamrher** [reply]
The STEND_CLUBB values at level 2 may be related to the limiters. I'd like to include some of the clubb stats diagnostics so we can see where the large STEND_CLUBB values are coming from in CLUBB. Ping me for those when you have time tmrw.

Could you also provide the path to the atm_in file?

[edit -- after thinking on this a bit more, the STEND_CLUBB bullseye's really could be anything; doesn't need to be the limiters. If RCMTEND_CLUBB is bad it would show up in STEND_CLUBB as well ... ]

**Reply by PeterHjortLauritzen** [reply]
sure

/glade/scratch/pel/clubb_hb_new_ne16pg3_ne16pg3_mg17_450_short/run/atm_in

**Reply by adamrher** [reply]
thanks. some relevant info -- prognostic momentum is off in this run and therefore u/v fields are not fed through the mfl limiter.

**Reply by vlarson** [reply]
Thanks, Peter.  0.0002Pa is a low pressure, compared to what CLUBB normally sees.  I wonder if [lines in CLUBB like these](https://github.com/larson-group/clubb_release/blob/91907b76e8af5973c6a6eb7cf120d5d6d24bae15/src/CLUBB_core/saturation.F90#L257-L261) are causing problems.  Does liquid water (or its tendency, RCMTEND_CLUBB) look reasonable at the top level?

**Reply by PeterHjortLauritzen** [reply]
I don't see any RCMTEND_CLUBB values larger than roundoff high up.

**Reply by vlarson** [reply]
I wonder where the energy (heat) is coming from if not latent heating.  It's possible that CLUBB's temperature monotonic flux limiter might be causing the problem.

**Comment by tto061** [comment]
Hi, I would also suggest again here (perhaps unnecessarily) to note and 
use ml2 as a tuning parameter.
It directly scales kvm and kvh.
I don't think there's anything magic about 1m or 30m for the HB mixing 
length.

Thomas Toniazzo
MISU          Stockholms university
SE-106 91 Stockholm         Sverige

**Comment by rgarc1a** [comment]
I have not commented on any of this because I have little to add. I understand the concern about HB and CLUBB interfering with each other. However, if HB can be implemented in the free atmosphere, or even just in the stratosphere and above, I believe it would be acceptable for WACCM.  HB attacks the most common form of instability in the free atmosphere: convective instability associated with large-amplitude resolved waves. It has worked well for that purpose for the last 20 years in the FV versions of WACCM; and Peter L. has shown it to also work for its intended purpose in SE.

**Comment by PeterHjortLauritzen** [comment]
Simulation for science evaluation of new HB code (free atmosphere HB mixing where CLUBB is not active) here (only difference between new run and baseline is the new HB code):

[ADF diags](https://webext.cgd.ucar.edu/FLTHIST/f.cam6_3_112.FLTHIST_v0c.ne30.non-ogw-ubcT-effgw0.3-rdg_beta0.5-vtrmi1.5-hb-tt.001/atm/) vs f.cam6_3_112.FLTHIST_v0c.ne30.non-ogw-ubcT-effgw0.3-rdg_beta0.5-vtrmi1.5.001

Summary:

- little impact on global metrics except (perhaps) SWCF:
  
<img width="428" alt="Screen Shot 2023-07-03 at 1 58 52 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/c0845ec6-7f60-40bd-b87b-69ee407d2bca">

<img width="1371" alt="Screen Shot 2023-07-03 at 1 59 25 PM" src="https://github.com/NCAR/amwg_dev/assets/8782989/ee94d118-415a-42f1-8f9e-4745923e4463">

Henceforth we will use the new HB code confguration for CAM7 simulations unless a better confguration "appears" (PR to be submitted this week).

**Comment by rgarc1a** [comment]
Thanks, Peter. That is overall good news. Any idea why it introduces changes in SWCF? Are the changes statistically significant?

**Comment by PeterHjortLauritzen** [comment]
Turning off the PBL part of HB led to this change in SWCF. 

Don't know about statistical significance.
